### PR TITLE
Migrate PVs and PVCs for pre v0.7.0 volumes

### PR DESCRIPTION
**Pull request by shuo-wu**
longhorn/longhorn#878

### CODE REVIEW

**Comment by yasker** [comment]
It's better we can run this as a separate tool rather than automatically. We want to user to know that something is changing their PV/PVC definitions. Also, it's unnecessary to add more complexity in the volume controller for this.

**Comment by shuo-wu** [comment]
How about creating a new API for the migration?

**Comment by yasker** [comment]
> How about creating a new API for the migration?

Why not create a separate tool? It can be a command from the manager. Why API is better?

File: app/migrate.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+	lhclientset "github.com/longhorn/longhorn-manager/k8s/pkg/client/clientset/versioned"
+)
+
+const (
+	FlagAll = "all"
+)
+
+func MigrateCmd() cli.Command {
+	return cli.Command{
+		Name: "migrate",
```
Comment (yasker) [inline app/migrate.go:29]: The name is not specific. Something like `migrate-pre-0.7.0-pv` would be more proper.

File: app/migrate.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+		}
+		if err = MigrateOldVolume(kubeClient, lhClient, lhNamespace, c.Args()[0]); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func MigrateOldVolume(kubeClient *kubeclientset.Clientset, lhClient *lhclientset.Clientset, lhNamespace, volumeName string) (err error) {
```
Comment (yasker) [inline app/migrate.go:88]: The name needs to be more specific.

File: app/migrate.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+	"github.com/longhorn/longhorn-manager/types"
+
+	lhclientset "github.com/longhorn/longhorn-manager/k8s/pkg/client/clientset/versioned"
+)
+
+const (
+	FlagAll = "all"
+)
+
+func MigrateCmd() cli.Command {
```
Comment (yasker) [inline app/migrate.go:27]: The name needs to be more specific.

File: app/migrate.go [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+		string(types.SettingNameDefaultLonghornStaticStorageClass), metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	if staticStorageClass.Value == "" {
+		return fmt.Errorf("empty static StorageClass for new PV/PVC creation")
+	}
+
+	if oldPV.Spec.PersistentVolumeReclaimPolicy != v1.PersistentVolumeReclaimRetain {
+		oldPV.Spec.PersistentVolumeReclaimPolicy = v1.PersistentVolumeReclaimRetain
```
Comment (yasker) [inline app/migrate.go:144]: PV's `retain` only refer to not delete the PV automatically when PVC deleted to my understanding, though I noticed we won't delete the volume from Longhorn as well. Can you point me to the code that skip the deletion? Just want to make sure it's safe to delete the PV.

Comment (shuo-wu) [reply app/migrate.go:144]: [This doc](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#retain) describes the behavior:
```
Retain:
Delete the PersistentVolume. The associated storage asset in external infrastructure (such as an AWS EBS, GCE PD, Azure Disk, or Cinder volume) still exists after the PV is deleted.

Delete:
For volume plugins that support the Delete reclaim policy, deletion removes both the PersistentVolume object from Kubernetes, as well as the associated storage asset in the external infrastructure, such as an AWS EBS, GCE PD, Azure Disk, or Cinder volume.
```

Comment (yasker) [reply app/migrate.go:144]: This looks good.

File: main.go [hunk 5]
```diff
@@ -48,6 +48,8 @@ func main() {
 		app.CSICommand(),
 		app.PostUpgradeCmd(),
 		app.UninstallCmd(),
+		// TODO: Remove MigrateForPre070VolumesCmd() after v0.8.0
```
Comment (yasker) [inline main.go:51]: `v0.8.1`

**Pull request merged**